### ISSUE TITLE
Support animated icons

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#13345]
We want to support animated icons, e.g like this menu-to-arrow animation:
![menu-arrow-animation](https://user-images.githubusercontent.com/1024117/33582722-76a82f2a-d90b-11e7-8e06-430975f93bab.gif)

We've considered a few possible approaches, taking into account that there isn't currently support for vector animation assets in Flutter(#1831):

1. Manually write code for the animation
2. Use animated image assets (e.g GIFs)
3. Use an icons font
4. Generate animation code from assets

While <i>#</i>1 (manual coding) is the fallback when all else fails, we hope to avoid it and use some form of an asset provided by designers.

We want to avoid <i>#</i>2 (GIFs) as it doesn't allow scaling and changing colors, and will require multiple assets for different screen densities.

## Icons font (no-go due to manual scrubbing)
I prototyped an icons font animation, it performs well, support color changes, and is not too bad with disk space (about 300 bytes for a frame for the sample menu->arrow animation).
A problem is that for icons like the menu->arrow it sometimes makes sense to let the user manually 'scrub' the animation progress and basically slow down the animation to an arbitrary speed. This doesn't work well with an animation that is based on a fixed number of frames.

## Generating animation code from assets (I'm currently working on this)
I prototyped generating Dart animation code from a sequence of SVG files for key frames, supporting only the minimal set of SVG features needed for the icon animation (which is currently: paths, a subset of the transitions, and opacity).

This is done assuming that the same paths exist in all frames, and has the same number of points and same IDs.

The generated code translates to ui.Canvas draw commands, and allows interpolating values and paths (by interpolating each path point).

To summarize the exploration:

- **(bad)** Requires a pre-processing stage.
- **(bad)** Supports a very limited subset of SVG.
- **(concerning)** If people start passing around "Dart animation asset" files which are essentially Dart code files it might be easy to hide malicious code in an asset.
- **(good)** Can directly use assets provided by designers (the designers need to be aware of the supported SVG subset, e.g make sure all shapes are converted to paths).
- **(good)** Can scale the animation to arbitrary size.
- **(good)** Can change the icon color from code.
- **(good)** Unused animation assets will be tree shaken and not take disk space in the shipping app.
- **(good)** We don't bake-in any vector format support into the framework (#1831 deserves proper homework)

**Comment by xqwzts** [comment]
There was some similar work being done in this library https://github.com/fabiomsr/lottie-flutter - it seems it got blocked by missing paint features in flutter.

**Comment by amirh** [comment]
Thanks @xqwzts,
There are still open questions around a vector drawable/animation support in the framework, mainly which format (#1831).
This is a smaller scoped task with the goal to enable animated icons in the framework.
I'm actually trying to get this done without baking any vector format support directly in the framework, to avoid premature decisions that will affect full vector support in the Flutter framework.
